### ISSUE TITLE
Autorotate gallery images

### ISSUE THREAD

**Issue by reagle** [saimn/sigal#72]
sigal seems smart enough to orient portrait images when it is within its own HTML framing. However, the actual jpgs (thumbnail and gallery images) are not rotated. I'd prefer the actual images be rotated so that if I want to use the image or thumbnail elsewhere on my website, it's in the correct orientation.

**Comment by reagle** [comment]
I'm wondering if there's been any movement on this one?

**Comment by saimn** [comment]
Not really, except another project which could be useful to save modified EXIF data (and based on Pillow): https://pypi.python.org/pypi/piexif

**Comment by reagle** [comment]
I haven't seen any progress at pillow [issue 520](https://github.com/python-pillow/Pillow/issues/520), any fixes available here?

**Comment by saimn** [comment]
Nothing new here ...

**Comment by jfleach** [comment]
Is there a workaround for this issue?

**Comment by saimn** [comment]
@jfleach - If you set only one of 'autorotate_images' and 'copy_exif_data', it should be ok: either rotate images and don't copy exif data, or copy exif and rely on browsers to do the rotation.

**Comment by anarcat** [comment]
In https://github.com/python-pillow/Pillow/issues/2800, Pil is considering adding https://pypi.python.org/pypi/piexif as a dependency, so it seems there is movement there.

**Comment by anarcat** [comment]
And as for browser-based rotation, it's also mostly stalled, as I documented elsewhere (https://github.com/Upload/Up1/issues/49#issuecomment-354450789). Basically, earlier CSS4 drafts had image orientation extensions that allowed the browser to automatically rotate (or not) images based on CSS directives, but that was pulled from the standard. It was argued this was better served in the semantic layer (HTML) but nothing has come up there.

So far only firefox has support for the deprecated CSS4 directive and there's a patch pending for Chrome, but it's been 4 years so I wouldn't hold my breath on that one, especially considering the time it takes for browser changes to make it across the whole ecosystem.

We need to rotate images on our own (losslessly, if I might add) *and* modify the related exif data. It's annoying gymnastics, but it's not really something we can avoid right now, regardless of what the up1 people say. ;)

(Incidentally, it might be nice to link to this bug report in the source code, sample config or warning messages so that people can follow progress more easily. I missed this bug report while filing #279 earlier.)

**Comment by saimn** [comment]
Great news if Piexif could be used in Pillow !

**Comment by saimn** [comment]
Pillow 6.0 added the possibility to modify the EXIF metadata and added `ImageOps.exif_transpose` :)
https://pillow.readthedocs.io/en/stable/releasenotes/6.0.0.html#added-imageops-exif-transpose

**Comment by reagle** [comment]
Any news?

**Comment by saimn** [comment]
Not really, the feature is in Pillow but time is lacking to update the code to use it.

**Comment by hjbaader** [comment]
I just looked at the code in this area. For me the images themselves get rotated correctly, but the thumbnails not. It looks like Transpose is missing from the thumbnail generation, is that correct?

**Comment by virtadpt** [comment]
That seems to be the case, I'm seeing the same thing over here.

**Comment by hjbaader** [comment]
Indeed. I added a parameter autorotate to generate_thumbnail and a passage to call Transpose in the same way as in generate_image and it seems to work.

**Comment by saimn** [comment]
This is another issue, but I have seen that recently too and committed a fix (a77b2e3f2e9d5ab2d79f6e2bd845dfc652f8aaf6).
That should happen only with `use_orig = True` because otherwise the thumbnails are created from the resized images which are rotated.
